src/System.Private.CoreLib/shared/Interop/Unix/System.Native/Interop.ReadDir.cs

```diff
@@ -30,16 +30,17 @@ internal unsafe struct DirectoryEntry
             internal byte* Name;
             internal int NameLength;
             internal NodeType InodeType;
-            internal const int NameBufferSize = 256;
+            internal const int NameBufferSize = 256; // sizeof(dirent->d_name) == NAME_MAX + 1
 
             internal ReadOnlySpan<char> GetName(Span<char> buffer)
             {
-                Debug.Assert(buffer.Length >= Encoding.UTF8.GetMaxCharCount(NameBufferSize - 1), "should have enough space for the max file name");
+                Debug.Assert(buffer.Length >= Encoding.UTF8.GetMaxCharCount(NameBufferSize - 1), $"should have enough space for the max file name, actual {buffer.Length} expected {Encoding.UTF8.GetMaxCharCount(NameBufferSize - 1)}");
```

Comment:
At first I removed the `-1` here. But `Encoding.UTF8.GetMaxCharCount(256)` gives 257 because in UTF8Encoding.cs:
```c#
            // Figure out our length, 1 char per input byte + 1 char if 1st byte is last byte of 4 byte surrogate pair
            long charCount = ((long)byteCount + 1);
```
If it's true that 256 bytes (with null) could produce 257 chars (with null) then we should expect callers to pass in a buffer of size 257.

However, I don't understand this comment. @tarekgh why would decoding a single byte that is the last byte of a 4 byte surrogate pair, lead to 2 characters?  (As opposed to last byte of a 3 byte surrogate pair, or last two bytes of a 4 byte surrogate pair, or something?)

Reply:
> why would decoding a single byte that is the last byte of a 4 byte surrogate pair, lead to 2 characters? 

This happens when you are using `System.Text.Decoder` and the decoder has 3 bytes cached. You decode one more byte and get 2 characters of output.

Reply:
In other words, `GetMaxCharCount` provides pessimistic estimate that works for any situation. You wish to have more accurate estimate for your given situation here and that's the problem.

Reply:
I'm fine with pessimistic. I just don't understand why it couldn't be the last byte of a two or three byte surrogate.
There will be no cached bytes in this context but I accept that the estimate can't know that. You are OK with me leaving as is, with the seemingly wrong -1?

Reply:
`NameBufferSize` includes null terminator. `buffer.Length` does not include the null terminator. So it should actually be `- 2` to be precise. If somebody looks at this code as is, they will likely think that the `- 1` is to componsate for the null terminator which is not the case.

I am not sure about the value of interpolated string. This string is going to be always formatted and it will make this method like 100x slower in debug builds. It is ok for debug builds to be slower, but not that much. 

I would change this to:

```
// No more than 1 char per byte
// buffer.Length does not include null terminator, NameBufferSize includes null terminator
Debug.Assert(buffer.Length >= NameBufferSize - 1);
```

Reply:
Yep, removed interopolated (although maybe `Debug.Assert(bool, string, string)` could have been OK
Changed to -2 with comment.